### ISSUE TITLE
PSA: onTriggerSend and onTriggerReceive will be deprecated.

### ISSUE THREAD

**Issue by SDGNelson** [SmartlyDressedGames/Unturned-3.x-Community#2432]
These events will be deprecated in the coming month as part of a rewrite to improve netcode performance. If your plugins are using them please comment here and we will find an alternative.

Rocket's usage of onTriggerSend has been replaced for the next update, and it did not use onTriggerReceive.

Edit: invoking RPCs by name will also be removed, and Rocket will be updated accordingly. Please comment where/why you are invoking RPCs manually.

**Comment by warren39** [comment]
Please add alternatives for onTriggerSend:
tellVehicleLock: For catching the locked or unlocked vehicle events. The `VehicleManager.onVehicleLockpicked` only work when Player using the item1353 to steal a vehicle.

tellVehicleExploded
tellVehicleDestroy

And also askUseObjectQuest  for onTriggerReceive, please.

**Comment by SDGNelson** [comment]
Thanks @warren39!

1. Do you want an event to be able to cancel the owner lock/unlock?
2. Will add exploded/destroy events.
3. Do you want to be able to cancel it, or just know before/after it happens?

**Comment by warren39** [comment]
> 1. Do you want an event to be able to cancel the owner lock/unlock?
Yes, It will be very easy to deny locking or unlocking vehicle.

> 3\. Do you want to be able to cancel it, or just know before/after it happens?
Just know after it happens. Thanks!

**Comment by SDGNelson** [comment]
An oversight on my part with this past update was plugins manually invoking RPCs by name. Invoking RPCs by name will also be removed. Any case where manually invoking the RPC is needed let me know and I will add an alternative.

**Comment by PandahutMiku** [comment]
> An oversight on my part with this past update was plugins manually invoking RPCs by name. Invoking RPCs by name will also be removed. Any case where manually invoking the RPC is needed let me know and I will add an alternative.

Hopefully, this won't be in the patch this afternoon, would cause more confusion on top of existing stuff.

For Clarification, you mean stuff like using SteamChannels to manually send packets?
ex.

`       ItemManager.instance.channel.send("tellTakeItem", ESteamCall.CLIENTS,
                                ESteamPacket.UPDATE_RELIABLE_BUFFER, xcord, ycord, instanceID);`

**Comment by SDGNelson** [comment]
@PandahutMiku Yes, not this afternoon. It is probably about a month away.

Regarding removing items: that can be added, but the regions will also be removed eventually, so I wonder where you are getting them from? i.e. Depending on the reasoning for deleting the item there might be a saner method to expose.

**Comment by PandahutMiku** [comment]
> @PandahutMiku Yes, not this afternoon. It is probably about a month away.
> 
> Regarding removing items: that can be added, but the regions will also be removed eventually, so I wonder where you are getting them from? i.e. Depending on the reasoning for deleting the item there might be a saner method to expose.

To prefix this, what is the intended purpose of removing manually invoking RPCs by name? There is a lot of different RPC's plugin devs use for a ton of various reasons. It might be more trouble than it's worth, but that's just my opinion.

Additionally, do you know when you are going to do the patch this afternoon for the server sending a messed up join packet that is making exit timers longer / disabling placements on vehicles (if going with default config?). It's relatively minor but not having vehicle placing is kind of annoying. Some people in Unturned's Community like Untured's Discord also believe that you will put out a magical update fixing all broken plugins, maybe it would be smart to put out a statement advising people to update their plugins.
~
~
~
~
~
I've listed all the uses of manual invoking RPC's we have right now below. I tried to double-check and make sure the same behavior wasn't already exposed somehow.

We use tellTakeItem to clear loot on the ground around the player in a configurable range. (i.e just going through all the Item Regions, finding the items nearby, removing them).


There isn't a clean way I know of, for a plugin to update the text on a sign. Same for a plugin opening/closing doors.  

We frequently need to clear entire player's inventories (gamemode plugins or just utils), and need to use tellSlot to update the player's equipped items. It looks like we could use sendSlot for that with the id of 0 and 1, but maybe the game should have a native function for clearing a player's inventory, as that stuff is pretty messy/probably subject to change in the future.

Garage Plugins are pretty popular and we have our own as well, where we have to use some reflection and the tellVehicle RPC in order to create a vehicle from scratch that includes all of the saved characteristics. (Could probably use a bunch of different methods to do it one RPC at a time)

We use tellVisualToggle from Player Clothing to manually toggle off a Player's Cosmetics for RP Events (so other players can clearly see they are Police), that's pretty niche though.

We use tellVehicleLock from VehicleManager to transfer a vehicle from one player to another. We use tellEnterVehicle to force people into vehicles.

We manually call tellChat as serverSendMessage from ChatManager enforces a 255 length cap, even though you can send and clients can receive longer chat messages. 255 characters aren't very much vs how large rich text messages can be sometimes.

Thanks for your time.

**Comment by SDGNelson** [comment]
@PandahutMiku Performance of invoking them without a name->index search will be a nice win. It might be possible to keep the existing send by name functionality for some, but not others (e.g. "chunk" RPCs replaced by NetPakWriter callback). It will be a ton of work on my end to upgrade them all too.

The patch should be within an hour I think, but I have been investigating misc reports. Just FYI an RSS feed and Discord channel for server updates was added recently: https://github.com/SmartlyDressedGames/U3-Docs/blob/master/ServerUpdateNotifications.md

Thanks for this list of RPCs! I will add maintained methods for all of them.

**Comment by Trojaner** [comment]
> Edit: invoking RPCs by name will also be removed, and Rocket will be updated accordingly. Please comment where/why you are invoking RPCs manually.

This will break a lot of plugins, and many of these are not actively maintained anymore. I believe deprecation would be the better solutin.

**Comment by SDGNelson** [comment]
@Trojaner I will deprecate where possible. My current experimental approach is to generate per-RPC methods for sending and receiving, so the deprecated equivalent would need to find the sending method and invoke it with the parameters casted to the appropriate types.

**Comment by SDGNelson** [comment]
@PandahutMiku addressing your points, here are the changes:
- `ItemManager.ServerClearItemsInSphere`
- `BarricadeManager.ServerSetSignText`
- `BarricadeManager.ServerSetDoorOpen`
- `VehicleManager.SpawnVehicleV3`
- `PlayerClothing.ServerSetVisualToggleState`
- `VehicleManager.ServerSetVehicleLock`
- Length exceptions removed from sendMessage trusting that plugins will use appropriate limits until variable height is added. Note anything longer than 2047 UTF8 bytes will be truncated.

Regarding the game modes I wonder if killing players is an option? Nervous to touch that code until I can get the chance to rewrite it properly.

**Comment by DiFFoZ** [comment]
In my plugins I'm using some RPC's.

1. For move barricade I'm using `BarricadeManager.tellTransformBarricade` and for structure `StructureManager.tellTransformStructure`.
2. Setting the pose of mannequin `BarricadeManager.tellPoseMannequin`.
3. `ResourceManager.tellResourceAlive` and `ResourceManager.tellResourceDead`.
4. `PlayerInventory.tellSize`.
5. I use `PlayerSkills.tellSkill` to use an alternative value of experience or escape experience decrease.
6. `PlayerClothing.tellSwapFace`.

**Comment by SDGNelson** [comment]
@DiFFoZ here are some changes:
1. `BarricadeManager.ServerSetBarricadeTransform` and `StructureManager.ServerSetStructureTransform`
2. `BarricadeManager.ServerSetMannequinPose`
3. `ResourceManager.ServerSetResourceAlive` and `ResourceManager.ServerSetResourceDead`
6. `PlayerClothing.ServerSetFace`

I wonder if you can elaborate on points 4 and 5? In particular for 5 there is already a config option for disabling skill loss, but maybe there is a more specific option needed.

**Comment by DiFFoZ** [comment]
For 4 I'm using this to the configurable size of inventory (e.g. Hands). 
And for 5:
- I'm using prefix patch `PlayerSkills.askUpgrade` to escape experience decrease or use instead of another value (e.g. uconomy).
- For commands, for example `/maxskills`. 

Also, maybe add an event OnSkillUpgrade with cancel upgrade.

**Comment by SDGNelson** [comment]
@DiFFoZ 
4. If you call inventory.items[2].resize(width, height) it should send the change to clients.
5. Adding `ServerUnlockAllSkills` and `OnSkillUpgraded_Global`

**Comment by Johnanater** [comment]
I use `AnimalManager.tellAnimal` for manually spawning a single animal (cursed, but it works)

I think that's the only one I use that hasn't been mentioned yet.

**Comment by SDGNelson** [comment]
@Johnanater hmm the downside of adding a public method for that would be plugin devs expecting it to work properly. 😛 Need to revamp the animals eventually.

**Comment by Luke254** [comment]
1. Event for `tellEquip`
2. Event before `tellHealth` is sent, because I want to edit health amount

**Comment by SDGNelson** [comment]
@Luke254 I will add an equipped event for plugins, but I wonder what you have in mind for modifying the health before send? There is a public `serverModifyHealth` method.

**Comment by Luke254** [comment]
> @Luke254 I will add an equipped event for plugins, but I wonder what you have in mind for modifying the health before send? There is a public `serverModifyHealth` method.

I made plugin which changes maximum health amount and I want to send correct percentage amount to player.

For example:
60/80 hp -> send 75 hp
45/150 hp -> send 30 hp
and so on

**Comment by SDGNelson** [comment]
@Luke254 hmm... for the modified max health it might be better to keep it as a patch or use one of the damage modifier events, as it would make more sense to have built-in max health support. I will do the equip event though.

**Comment by rube200** [comment]
Is there any alternative to force rain or stop it??
before this update i used `tellLightingRain`

**Comment by SDGNelson** [comment]
@rube200 `ResetScheduledWeather` to stop, `ForecastWeatherImmediately` to start now according to level, `SetScheduledWeather` for manual scheduling, or `SetPerpetualWeather` for permanent

**Comment by Pustalorc** [comment]
I'm currently using the following RPCs:
`tellToggleGenerator`
`tellToggleFire`
`tellToggleOven`
`tellToggleOxygenator`
`tellToggleSafezone`
`tellToggleSpot`
`tellUpdateStereoTrack`
All of them are used on AutoTurnOff, which turns all of those interactables from the on state to the off state when a player disconnects (no matter where they are in the map), and there's no built in method for that that does not check the player that called the server-sided RPC.

**Comment by CyberAndrii** [comment]
Need a method to despawn item instead of [using `ItemManager.tellTakeItem`](https://github.com/openmod/openmod/blob/bdf090b244096d6f7183f42a4356f2ef3e48e96f/unturned/OpenMod.Unturned/Items/UnturnedItemDrop.cs#L68).

**Comment by SDGNelson** [comment]
@Pustalorc Ok, will do.
@CyberAndrii Where are you getting the item data from? Asked a similar question to Miku which is why ItemManager.ServerClearItemsInSphere was added.

**Comment by CyberAndrii** [comment]
> Where are you getting the item data from?

- By iterating over `ItemManager.regions.items` ([see](https://github.com/openmod/openmod/blob/cb307bfd1726c5ad8af4fb0ca35b4237efd67419/unturned/OpenMod.Unturned/Items/UnturnedItemDirectory.cs#L28..L31)) to get all dropped items.
- By calling `ItemManager.dropItem` and then searching for the item data in `ItemManager.regions[x, y]` ([see](https://github.com/openmod/openmod/blob/d64e9bf7bf1fb053b49577d4c641ce6850d7f2d7/unturned/OpenMod.Unturned/Items/UnturnedItemSpawner.cs#L149..L157))

**Comment by SDGNelson** [comment]
@CyberAndrii to clarify, what is the purpose there? e.g. some custom despawn timer? Because I want to remove the "regions" code eventually.

**Comment by CyberAndrii** [comment]
To provide plugins a universal API (spawn, despawn, return a list of all currently spawned items) so they can be used in multiple games without depending on them. Because of abstractions, I need a way to despawn only a single item at a time. I don't know how exactly plugins will use that.

Also, would be fine if you add methods that will accept only instance id (or something else?) instead of regions.

**Comment by SDGNelson** [comment]
@CyberAndrii using instance ID sounds reasonable. Though I will need to add an instance ID to region lookup. Since it is a multi-game abstraction do you know if anyone is using it for anything at the moment?

**Comment by al3xanderwalker** [comment]
I'm currently using `ItemManager.tellTakeItem` using the x and y provided from patching `ItemManager.askTakeItem` is there any other way to clear a specific item using  information from it?

**Comment by al3xanderwalker** [comment]
Also is there any alternative for `UseableGun.askReload` and `PlayerCrafting.tellCraft`?
I use different logic for reloading and so need to ask them to do the reload animation at the end using `askReload`.
Same for `tellCraft`, I use different crafting logic however it doesn't update the resource count in their inventory after its completion without sending that call.

**Comment by SDGNelson** [comment]
@al3xanderwalker What are you patching in askTakeItem? And what does your custom reload and crafting do?

**Comment by PandahutMiku** [comment]
As of this morning, we've been using all built-in methods and no invoking RPC's by name. I didn't notice any real issues during my testing and none have been reported either, all the replacement methods sent seem to be working great.

It would be nice if we had a method in PlayerStance for the server to set stance. Right now, for example, with PlayerStance, the server can use checkStance to try to force a new player stance but it checks for cooldown expiring and a bunch of other checks. Just want a clean way to force a player back into a stance if required by plugin, ignoring cooldown.

It might be nice to have one in PlayerAnimator too, I believe sendGesture works fine/has no cooldowns, but not sure if you want us to be using that.

**Comment by SDGNelson** [comment]
@PandahutMiku Glad to hear that! The reason I encapsulated stance changes was to help prevent the client and server from getting out of sync. I wonder what sort of stance requirement plugins you have in mind?

**Comment by PandahutMiku** [comment]
> @PandahutMiku Glad to hear that! The reason I encapsulated stance changes was to help prevent the client and server from getting out of sync. I wonder what sort of stance requirement plugins you have in mind?

For example, we have a downed/revive plugin, that when a player hits 0 HP, instead of dying, they are placed into a downed state in which they are forced into Prone/unable to equip weapons/surrender start. When the player tries to stand up, we can patch the stance methods server-side and prevent the stance change from replicating to other clients, but we also want to inform the client themselves that they are not standing up (since stance is client-side/doesn't depend on server to confirm). If we just use checkStance the second they try to stand up, we hit the cooldown.

**Comment by SDGNelson** [comment]
@PandahutMiku Hmmm... my first thought was to suggest an event for cancelling prone/crouch requests, but that does not solve the issue of forcing into prone if the timer is active when the player dies. Instead how about a stance property for the server to override the player's input so that the other stance rules are followed? e.g. Player is pressing stand up, but server has a "wants to prone" override which takes priority.

**Comment by PandahutMiku** [comment]
> @PandahutMiku Hmmm... my first thought was to suggest an event for cancelling prone/crouch requests, but that does not solve the issue of forcing into prone if the timer is active when the player dies. Instead how about a stance property for the server to override the player's input so that the other stance rules are followed? e.g. Player is pressing stand up, but server has a "wants to prone" override which takes priority.

That would work and be a lot better than just spamming the client with prone requests whenever they try to change stance.

**Comment by al3xanderwalker** [comment]
@SDGNelson It's kind of hard to explain but I've been making it so items can be stacked in your inventory e.g. 200x logs, askTakeItem is patched so that if they attempt to pickup an item, it will try and stack it with others first before adding it to a new stack. The custom reload rather than loading a new magazine into the gun, it takes ammo from an existing magazine, reducing that and increasing the magazine in the guns amount, the reload animation is then called after as the original logic is overridden. The crafting is modified to allow for using stacked items and so it must be cancelled as the normal crafting logic throws an error if left the continue.

**Comment by SDGNelson** [comment]
@PandahutMiku Indeed. Okay, I have opened a separate issue (#2470) for that request.

@al3xanderwalker Impressive that you managed to get that working. I don't think I can support those officially yet, but in particular better per-item quantity/amount/stacking features are planned for the future.

**Comment by al3xanderwalker** [comment]
It works to a surprisingly effective degree, the only issues being the inability to tell the player to do the reload animation or update the resources in the crafting UI, is there any plans for replacement methods for these?

**Comment by Jerico97** [comment]
@SDGNelson Hello! 
I'm need next alternatives:
1. askReload
2. askAimStart\askAimStop
3. tellInspect (I need the moment of playing the animation of the inspection of the weapon)
Thanksv

**Comment by educatalan02** [comment]
I am using this ones: tellVehicleHealth, tellVehicleLock, tellVehicleFuel, tellVehicleBatteryCharge, tellBarricade, tellExitVehicle, tellVehicleDestroy, tellToggleDoor

**Comment by SDGNelson** [comment]
@al3xanderwalker I suppose I can add methods for those, but patching into them may require changes with the rewrites.

@Jerico97 Okay, I will add events for those.

@educatalan02: Okay, I will add events for those.

**Comment by IAmSilK** [comment]
Will instances where we patch `tell...` methods result in plugins breaking? Or is it simply `SteamChannel.send` being removed?

**Comment by SDGNelson** [comment]
@IAmSilK unfortunately they will break patching in most cases because the CSteamID senderId parameter will be removed, so having officially supported events is preferred. That being said it should be straightforward for maintained plugins to update.

**Comment by IAmSilK** [comment]
So methods which take said `CSteamID sender` parameter won't be able to called. How about `ask...` methods without the `CSteamID` such as `VehicleManager.askVehicleDestroy`.

**Comment by SDGNelson** [comment]
@IAmSilK All of them will still be callable.

**Comment by IAmSilK** [comment]
Sounds good. Here's my list of stuff used, sorry if some of these have already been mentioned.

For `SteamChannel.send`, I'm currently using the following:

- `tellBarricadeUpdateState` for individual players. I use this to make players believe they can interact with items so it triggers a patch server side when they attempt to. `checkToggle` or similar methods are then patched so a plugin can dictate whether or not something can be interacted with.
- `tellBarricadeOwnerAndGroup` for individual players. Same reason as above.
- `tellRevive` allows for respawning players at a custom position.
- `tellSetGroup` allows for more control over player's groups. For example: creating groups where there is no owner as they are controlled by a plugin. There may already be an alternative to this.
- `tellEffectClearByID` for all players would be nice, but not necessary.
- `tellStance` there might already be a better way to do this.

For patching of server side RPCs (in need of events):

- `tellRevive` - intercepts revives then custom respawn positions can be set.
- `tellEquip` - an event for after an item is equipped.
- `tellEnterVehicle` - an event for after a player enters a vehicle (onEnterRequested can be cancelled)
- `sendExitVehicle` - an event for after a player exits a vehicle. This may not be removed, but I thought I'd add it just in case.

**Comment by rube200** [comment]
On my side i only need:
`tellClaimBed`
`tellClothing`

**Comment by SDGNelson** [comment]
Note these will be on the network beta branch when that goes up later this week, but not in the Friday update because with all these changes these are causing merge conflicts.

@al3xanderwalker:
- Adding `ServerPlayReload` and `ServerRefreshOwnerCrafting`.

@Jerico97:
- Adding `OnReloading_Global` for askReload and `OnAimingChanged_Global` for askAimStart and askAimStop.
- Adding `OnInspectingUseable_Global` for tellInspect.

@educatalan02:
- Adding `OnHealthChanged_Global`, `OnLockChanged_Global`, `OnFuelChanged_Global`, and `OnBatteryLevelChanged_Global`.
- Adding `OnDoorChanged_Global` for tellToggleDoor.
- For tellBarricade I wonder if `onBarricadeSpawned` does what you need?
- For tellDestroyVehicle `OnPreDestroyVehicle` should do the job.
- Adding `OnPassengerAdded_Global`, `OnPassengerChangedSeats_Global`, and `OnPassengerRemoved_Global`.

@IAmSilk:
- I don't think I can support per-player barricade state officially yet because it breaks a lot of the base game assumptions, but it is something that will need to be added sooner or later.
- Adding `OnSelectingRespawnPoint` event for custom spawns.
- For groups are you using the in-game "dynamic groups", or treating them as Steam clans?
- Adding `ClearEffectByID_AllPlayers`.
- To change stance you can set the `stance` property which calls `checkStance`, or directly call `checkStance`.
- In the most recent update an `OnUseableChanged_Global` was added for `tellEquip`.
- Adding `OnPassengerAdded_Global`, `OnPassengerChangedSeats_Global`, and `OnPassengerRemoved_Global`.

@rube200:
Okay, I have added `ServerClaimBedForPlayer` and `ServerUnclaimBed`. Can you elaborate on what you are using tellClothing for please? It will need to be reworked because it has so many parameters.

**Comment by IAmSilK** [comment]
Firstly, thanks for all the support your providing for plugin devs through this big change.

> I don't think I can support per-player barricade state officially yet because it breaks a lot of the base game assumptions, but it is something that will need to be added sooner or later.
Are there any examples of the new workings of the net code? I'm wondering if it'll be possible to unofficially patch the new sending mechanics to modify the state per player.

> For groups are you using the in-game "dynamic groups", or treating them as Steam clans?
I am using dynamic groups.

**Comment by rube200** [comment]
> @rube200:
> Okay, I have added `ServerClaimBedForPlayer` and `ServerUnclaimBed`. Can you elaborate on what you are using tellClothing for please? It will need to be reworked because it has so many parameters.

I am currently using it to save / load the player's inventory from the database.
If you’re going to rework it, I’ll wait for it.
Atm my plugin saves the player's inventory when disconnect and when player connects back the plugin start to get information from the database, then when server start receiving askClothing calls my plugin responds to them manually.

**Comment by SDGNelson** [comment]
@IAmSilK not on the preview branch yet, but it should be soon this week. If you update your code it should be possible to continue doing what you were doing. Will check the dynamic groups this afternoon.

@rube200 to clarify: the functionality will be identical. Yeah the save/load sucks without proper support for async requests... I want to revamp the save system eventually™. The once-per-player RPCs e.g. askClothing are also going to be removed, but once that is done it should be safe to override the items immediately upon receipt.

**Comment by SDGNelson** [comment]
For anyone interested an early version with the general changes is up on the preview branch now. More details  #2479

**Comment by IAmSilK** [comment]
Looking at the new code, will it be safe to relate an instance of `ITransportConnection` back to its player by calling something like `Provider.clients.FirstOrDefault(player => player?.steamPlayer != null && player.steamPlayer.transportConnection == transportConnection)`?

Due to no official barricade state per player events, I'm looking at patching `BarricadeManager.askBarricades(ITransportConnection  ...)` and need a way to relate `ITransportConnection` to its owner player.

**Comment by SDGNelson** [comment]
@IAmSilK that should work, or there is a method `Provider.findPlayer` which takes a transport connection parameter.

Since the caller has that information I will rename `askBarricades` to `SendRegion`, and swap out the transport connection parameter for player.

**Comment by robelkint** [comment]
Hey Nelson,

I would just need an alternative for `askVoiceChat`.
I just need it to grab the byte array in the parameters to deal with as an audioclip.

Cheers

**Comment by SDGNelson** [comment]
@robelkint what do you use the voice payload for? Is it a clientside plugin? Because I cannot imagine it is much use without the Steam client decoding.

**Comment by robelkint** [comment]
It'd be serverside. Basically I'd like to grab the info and try and play the audio at another source. Sort of like the walkie talkies but without the receiving player.

**Comment by SDGNelson** [comment]
@robelkint that would not be possible without vanilla support for non-player voice chat audio sources, sorry. Would be interesting to add tunable radios for that down the line.

**Comment by robelkint** [comment]
Ahhh okay, appreciate the feedback though! That was essentially the idea there. Would be cool to see implemented at some point later on.

**Comment by DiFFoZ** [comment]
Is there's has alternative for `VehicleManager.tellVehicleEnter`? 
Need a method something like `VehicleManager.forceEnterPlayer` that doesn't check if the player is looking at the vehicle, is the vehicle too far, etc.

**Comment by SDGNelson** [comment]
@DiFFoZ okay, I will add a method for that.

**Comment by SDGNelson** [comment]
Closing since this has cooled down and the RPC update is nearly ready. Thank you all for updating your code! 🙂

**Comment by SDGNelson** [comment]
Thanks again for working with me to update! This will be releasing today in the 3.21.13.0 update.
